README.md
SimilarTracks for Kodi
===========

A script that uses the currently playing track to generate a playlist of similar tracks in your Kodi music library.

Features
--------
- Create a playlist of tracks from your music library that are similar to the currently playing track.
- Uses Last.fm getSimilar API to find similar tracks.
- Runs in background.
- Can specify the maximum number of tracks to add to the playlist.

Screenshots
----------

<img alt="Get similar tracks progress dialog" src="https://raw.github.com/brianhornsby/www_brianhornsby_com/master/img/similartracks_get_dialog.png" height="128"/>
<img alt="Found similar tracks progress dialog" src="https://raw.github.com/brianhornsby/www_brianhornsby_com/master/img/similartracks_found_dialog.png" height="128"/>
<img alt="Added tracks to playlist progress dialog" src="https://raw.github.com/brianhornsby/www_brianhornsby_com/master/img/similartracks_added_dialog.png" height="128"/>
<img alt="Get similar tracks notification" src="https://raw.github.com/brianhornsby/www_brianhornsby_com/master/img/similartracks_get_notification.png" height="128"/>
<img alt="Added tracks to playlist notification" src="https://raw.github.com/brianhornsby/www_brianhornsby_com/master/img/similartracks_added_notification.png" height="128"/>

Installation
-----------
Download the latest zip file and install the addon. See http://kodi.wiki/view/HOW-TO:Install_an_Add-on_from_a_zip_file for more details on installing addons from zip file.

Usage
-----
The SimilarTracks script can be accessed from the Programs menu or called using the RunScript builtin function (RunScript(script.similartracks))

Settings
--------
The following settings are available.

**Maximum tracks to add to playlist**: The maximum tracks that will be added to the playlist. Default: 25

**Run in background**: Script runs in the background. Default: false

**Playlist order**: The order that tracks should be added to the playlist. Random: Tracks are added at random. Last.FM: Tracks are added in the order they came back from Last.fm. Default: Random

License
-------
SimilarTracks for Kodi is licensed under the [GPL 3.0 license] [1].


[1]: http://www.gnu.org/licenses/gpl-3.0.html


changelog.txt
*** 3.0.0 ***
Updated plugin to use Kodi name. Updated copyright date. [d9ac0ea]

*** 2.0.1 ***
Fix for issue #1 where unicode character was causing script to raise an exception.

*** 2.0.0 ***
Complete rewrite of code.

*** 1.1.2 ***
Updated JSON calls to work with Frodo.

*** 1.1.1 ***
Updated code to pep8 coding style.
Fixed script error if Last.fm returned zero records.

*** 1.1.0 ***
Added 'run in background' setting.
Fixed 'maximum count' setting.

*** 1.0.1 ***
Fixed 'UnicodeDecodeError' exception.

*** 1.0.0 ***
Updated version number to 1.0.0.
Set extension point to script rather than lyric.

*** 0.3.0 ***
Left debugging statement in code.

*** 0.2.0 ***
Changed XBMC music library lookup to use json RPC.

*** 0.1.0 ***

Changed to lyric script.
Tidied up code.

*** 0.0.1 ***

Initial release.


similartracks.py
# main imports
import sys
import os
import xbmc
import xbmcgui
import xbmcplugin
import xbmcaddon
import array
import urllib
from urllib2 import urlopen
import xml.dom
from xml.dom import minidom
import re

pDialog = xbmcgui.DialogProgress()
pDialog.create('Similar Tracks', 'Initializing script...')

__settings__      = xbmcaddon.Addon(id='script.similartracks')
__language__      = __settings__.getLocalizedString
__version__       = __settings__.getAddonInfo('version')
__cwd__           = __settings__.getAddonInfo('path')
__maxcount__      = __settings__.getSetting('maxcount')

__lastfmtracks__ = {}
__playlisttracks__ = {}

def get_next_track_to_add(index, prev_artist):
   for i in __playlisttracks__:
      if __playlisttracks__[i]["artist"] != prev_artist:
         if __playlisttracks__[i]["added"] == False:
            return i
   for i in __playlisttracks__:
      if __playlisttracks__[i]["added"] == False:
         return i
   return -1

def get_track_index_in_playlist(filename):
   for index in __playlisttracks__:
      if __playlisttracks__[index]["filename"] == filename:
         return index
   return 0

def get_lastfm_similar_tracks(artist, track):
   url = "http://ws.audioscrobbler.com/2.0/?method=track.getsimilar&artist=" + urllib.quote_plus(artist) + "&track=" + urllib.quote_plus(track) + "&api_key=b25b959554ed76058ac220b7b2e0a026"
   xmlhandle = urlopen(url)
   xmldoc = minidom.parse(xmlhandle)
   xmlhandle.close()

   trackcount = 0
   __lastfmtracks__.clear()

   pDialog.update(50, "Processing similar tracks...", "", "")

   for node in xmldoc.childNodes:
      if node.nodeName == "lfm":
         for node in node.childNodes:
            if node.nodeName == "similartracks":
               for node in node.childNodes:
                  if node.nodeName == "track":
                     track = {}
                     for node in node.childNodes:
                        if node.nodeName == "name":
                           track["title"] = node.firstChild.nodeValue
                        elif node.nodeName == "artist":
                           for node in node.childNodes:
                              if node.nodeName == "name":
                                 track["artist"] = node.firstChild.nodeValue
                     __lastfmtracks__[trackcount] = track
                     trackcount = trackcount + 1

def get_similar_tracks(artist, title):
   get_lastfm_similar_tracks(artist, title)
   pDialog.update(75, "Creating playlist...", "Received %d similar tracks to" % (len(__lastfmtracks__)), "%s - %s" % (artist, title))
   count = 0

   for index in __lastfmtracks__:
      if count >= int(__maxcount__):
         break

      track = __lastfmtracks__[index]
      sql_music = "select distinct strTitle, strArtist, strFilename, strPath from songview where idSong = (select idSong from song where strTitle = \'%s\') and idArtist = (select idArtist from artist where strArtist = '%s')" % ( track["title"].replace("'", "''"), track["artist"].replace("'", "''"))
      music_xml = xbmc.executehttpapi( "QueryMusicDatabase(%s)" % urllib.quote_plus( sql_music.encode('utf-8')), )
      if len(music_xml) > 0 and (not music_xml.startswith("<li>Error")):
         fields = re.findall( "<field>(.*?)</field>", music_xml, re.DOTALL )
         if len(fields) == 4:
            path = fields[3] + fields[2]
            if path != currenttrack and get_track_index_in_playlist(path) == 0:
               track = {}
               track["title"] = fields[0]
               track["artist"] = fields[1]
               track["filename"] = path
               track["added"] = False
               __playlisttracks__[count] = track
               count = count + 1
               pDialog.update(85, "Creating playlist. Adding %s" % fields[1])
   return count

tag = xbmc.Player().getMusicInfoTag()
if xbmc.Player().isPlayingAudio():            
   playlist = xbmc.PlayList(0)
   currenttrackpos = playlist.getposition()
   currenttrack = playlist[currenttrackpos].getfilename()
   pDialog.update(25, "Getting similar tracks for %s - %s" % (tag.getArtist(), tag.getTitle()), "")
   count = get_similar_tracks(tag.getArtist(), tag.getTitle())

   print "Found %d similar tracks" % count

   if count > 0:
      trackpos = currenttrackpos + 1
      while xbmc.PlayList(0).size() > trackpos:
         xbmc.PlayList(0).remove(xbmc.PlayList(0)[trackpos].getfilename())

      index = 0
      playlist_len = len(__playlisttracks__)
      previous_artist = tag.getArtist()
      while index < playlist_len:
         i = get_next_track_to_add(index, previous_artist)
         if i == -1:
            break
         previous_artist = __playlisttracks__[i]["artist"]
         listitem = xbmcgui.ListItem(__playlisttracks__[i]["title"])
         xbmc.PlayList(0).add(__playlisttracks__[i]["filename"], listitem)
         __playlisttracks__[i]["added"] = True
         index = index + 1
  
   pDialog.close()

   if count == 0:
      dialog = xbmcgui.Dialog()
      ok = dialog.ok('Similar Tracks', 'Unable to find any similar tracks for',  '%s - %s' % (tag.getArtist(), tag.getTitle()))


resources/lib/kodiutils.py
#/*
# *
# * SimilarTracks for Kodi.
# *
# * Copyright (C) 2015 Brian Hornsby
# *
# * This program is free software: you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation, either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program.  If not, see <http://www.gnu.org/licenses/>.
# *
# */

import xbmc
import xbmcgui
import unicodedata
import urllib


def normalize_unicode(text):
    if text and not isinstance(text, unicode):
        return text
    if not text or len(text) == 0:
        return ''
    return unicodedata.normalize('NFKD', text).encode('ascii', 'ignore')


def check_value(value):
    if value is None:
        return ''
    return normalize_unicode(value)


def get_value(tuple, key):
    if key not in tuple:
        return ''
    return check_value(tuple[key])


def check_int(value):
    if value is None:
        return 0
    return value


def get_int(tuple, key):
    if key not in tuple:
        return 0
    return int(check_value(tuple[key]))


def add_params(root, params):
    return '%s?%s' % (root, urllib.urlencode(params))


def keyboard(default='', heading='', hidden=False):
    kb = xbmc.Keyboard(default, heading, hidden)
    kb.doModal()
    if (kb.isConfirmed() and len(kb.getText()) > 0):
        return kb.getText()
    return None


def notification(header, message, time=5000, image=''):
    command = 'Notification(%s, %s, %s, %s)' % (header, message, time, image)
    xbmc.executebuiltin(command)


def ok(heading, line1, line2='', line3=''):
    dialog = xbmcgui.Dialog()
    dialog.ok(heading, line1, line2, line3)


def select(heading, list):
    dialog = xbmcgui.Dialog()
    return dialog.select(heading, list)


def yesno(heading, line1, line2='', line3=''):
    dialog = xbmcgui.Dialog()
    return dialog.yesno(heading, line1, line2, line3) == 1


def get_params(text):
    param = []
    paramstring = text
    if (len(paramstring) >= 2):
        params = text
        cleanedparams = params.replace('?', '')
        if (params[len(params) - 1] == '/'):
            params = params[0:len(params) - 2]
        pairsofparams = cleanedparams.split('&')
        param = {}
        for i in range(len(pairsofparams)):
            splitparams = {}
            splitparams = pairsofparams[i].split('=')
            if len(splitparams) == 2:
                param[splitparams[0]] = urllib.unquote_plus(splitparams[1])
    return param


resources/lib/kodisettings.py
#/*
# *
# * SimilarTracks for Kodi.
# *
# * Copyright (C) 2015 Brian Hornsby
# *
# * This program is free software: you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation, either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program.  If not, see <http://www.gnu.org/licenses/>.
# *
# */

import xbmcaddon
import xbmc


class KodiSettings:
    def __init__(self, addonid, argv):
        self.__addon__ = xbmcaddon.Addon(id=addonid)
        self.__argv__ = argv
        self.__id__ = self.__addon__.getAddonInfo('id')
        self.__datapath__ = 'special://profile/addon_data/%s' % (self.__id__)
        self.__name__ = self.__addon__.getAddonInfo('name')
        self.__path__ = self.__addon__.getAddonInfo('path')
        self.__version__ = self.__addon__.getAddonInfo('version')

    def __getitem__(self, key):
        value = self.__addon__.getSetting(key)
        if value.isdigit():
            return int(value)
        else:
            return value

    def get_string(self, id):
        return self.__addon__.getLocalizedString(id)

    def get_argv(self, idx):
        return self.__argv__[idx]

    def get_argc(self):
        return len(self.__argv__)

    def get_datapath(self, path=''):
        return xbmc.translatePath('%s/%s' % (self.__datapath__, path))

    def get_name(self):
        return self.__name__

    def get_path(self, path=''):
        return xbmc.translatePath('%s/%s' % (self.__path__, path))

    def get_version(self):
        return self.__version__

    def open(self):
        self.__addon__.openSettings()


resources/lib/xbmcsettings.py
#/*
# *
# * SimilarTracks add-on for XBMC.
# *
# * Copyright (C) 2012 Brian Hornsby
# *	
# * This program is free software: you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation, either version 3 of the License, or
# * (at your option) any later version.
# *	
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *	
# * You should have received a copy of the GNU General Public License
# * along with this program.  If not, see <http://www.gnu.org/licenses/>.
# *
# */

import xbmcaddon, xbmc

class Settings:
	def __init__(self, addonid, argv):
		self.__addon__ = xbmcaddon.Addon(id=addonid)
		self.__argv__ = argv
		self.__id__ =  self.__addon__.getAddonInfo('id')
		self.__datapath__ = 'special://profile/addon_data/%s' % (self.__id__)
		self.__name__ = self.__addon__.getAddonInfo('name')
		self.__path__ = self.__addon__.getAddonInfo('path')
		self.__version__ = self.__addon__.getAddonInfo('version')

	def __getitem__(self, key):
		value = self.__addon__.getSetting(key)
		if value.isdigit():
			return int(value)
		else:
			return value
	
	def get_string(self, id):
		return self.__addon__.getLocalizedString(id)

	def __getitem__(self, key):
		value = self.__addon__.getSetting(key)
		if value.isdigit():
			return int(value)
		else:
			return value

	def get_argv(self, idx):
		return self.__argv__[idx]

	def get_datapath(self, path=''):
		return xbmc.translatePath('%s/%s' % (self.__datapath__, path))

	def get_name(self):
		return self.__name__

	def get_path(self, path=''):
		return xbmc.translatePath('%s/%s' % (self.__path__, path))

	def get_version(self):
		return self.__version__

	def open(self):
		self.__addon__.openSettings()


default.py

#/*
# *
# * SimilarTracks for Kodi.
# *
# * Copyright (C) 2015 Brian Hornsby
# *
# * This program is free software: you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation, either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program.  If not, see <http://www.gnu.org/licenses/>.
# *
# */

import random
import sys
import urllib
import xbmc
import xbmcgui
from urllib2 import urlopen, URLError

import resources.lib.kodisettings as kodisettings
import resources.lib.kodiutils as utils

if sys.version_info < (2, 7):
    import simplejson
else:
    import json as simplejson

_addonid = 'script.similartracks'
_settings = kodisettings.KodiSettings(_addonid, sys.argv)

# Get addon information and settings.
_addonname = _settings.get_name()
_version = _settings.get_version()
_maxcount = _settings['maxcount']
if not _maxcount:
    _maxcount = -1
_runinbackground = (_settings['background'] == 'true')
_order = int(_settings['order'])


def log_debug(msg):
    if _settings['debug'] == 'true':
        print '%s: DEBUG: %s' % (_addonid, msg)


def log_error(msg):
    print '%s: ERROR: %s' % (_addonid, msg)

log_debug('Addon Id:   [%s]' % (_addonid))
log_debug('Addon Name: [%s]' % (_addonname))
log_debug('Version:    [%s]' % (_version))

log_debug('maxcount: %d' % _maxcount)
log_debug('runinbackground: %d' % _runinbackground)
log_debug('order: %d' % _order)

if not _runinbackground:
    pDialog = xbmcgui.DialogProgress()
    pDialog.create(
        _settings.get_string(1000), _settings.get_string(3000))


def display_notification(header, message):
    image = _settings.get_path('icon.png')
    utils.notification(header, message, image=image)


def get_lastfm_similar_tracks(artist, track):
    base_url = 'http://ws.audioscrobbler.com/2.0/?'
    params = {'method': 'track.getsimilar', 'artist': artist, 'track':
              track, 'api_key': '5da513b631898f5372a5e5f863651212', 'format': 'json',
              'autocorrect': 1}
    url = '%s%s' % (base_url, urllib.urlencode(params))
    log_debug('Last.fm URL: %s' % url)

    try:
        f = urlopen(url)
    except URLError as exception:
        log_error(exception)
        utils.ok(_settings.get_string(1000), _settings.get_string(3007))
        return []

    json_query = unicode(f.read(), 'ascii', errors='ignore')
    f.close()
    json_response = simplejson.loads(json_query)
    lastfmtracks = []
    if 'similartracks' in json_response:
        for track in json_response['similartracks']['track']:
            lastfmtracks.append({'title': track[
                                'name'], 'artist': track['artist']['name']})
    return lastfmtracks


def get_similar_tracks(artist, title):
    log_debug('Looking for similar tracks to %s - %s' % (artist, title))
    lastfmtracks = get_lastfm_similar_tracks(artist, title)
    message = (_settings.get_string(3003) % (len(lastfmtracks)))
    if _runinbackground:
        display_notification(_settings.get_string(1000), message)
    else:
        pDialog.update(75, _settings.get_string(3002), message)
    log_debug('Last.fm returned %d similar tracks' % (len(lastfmtracks)))

    if not _runinbackground and pDialog.iscanceled():
        return (0, [])

    json_query = xbmc.executeJSONRPC(
        '{"jsonrpc": "2.0", "method": "AudioLibrary.GetArtists", "params": {"properties": [], "sort": { "method": "label" } }, "id": 1}')
    json_query = unicode(json_query, 'ascii', errors='ignore')
    json_response = simplejson.loads(json_query)

    artists = []
    if (json_response['result'] is not None) and ('artists' in json_response['result']):
        for artist in json_response['result']['artists']:
            artists.append(
                {'artist': artist['artist'], 'id': artist['artistid']})

    count = 0
    playlisttracks = []
    for track in lastfmtracks:
        tracktitle = track['title'].encode('ascii', 'ignore')
        trackartist = track['artist'].encode('ascii', 'ignore')
        artistid = None
        for artist in artists:
            if isinstance(artist['artist'], str) or isinstance(artist['artist'], unicode):
                if 'artist' in artist and artist['artist'].encode('ascii', 'ignore') == trackartist:
                    artistid = artist['id']
                    break
        if artistid:
            json_query = xbmc.executeJSONRPC(
                '{"jsonrpc": "2.0", "method": "AudioLibrary.GetSongs", "params": {"properties": ["title", "artist"], "sort": { "method": "label" },  "filter": {"artistid": %s} }, "id": 1}' % artistid)
            json_query = unicode(json_query, 'ascii', errors='ignore')
            json_response = simplejson.loads(json_query)
            if (json_response['result'] is not None) and ('songs' in json_response['result']):
                for song in json_response['result']['songs']:
                    if 'title' in song and song['title'] == tracktitle:
                        playlisttracks.append({'songid': song[
                                               'songid'], 'artist': song['artist'], 'title': song['title']})
                        count = count + 1
                        if not _runinbackground:
                            pDialog.update(85, _settings.get_string(3004) % ('%s - %s' % (
                                trackartist, tracktitle)), _settings.get_string(3001) % count)
                        break

    return (count, playlisttracks)


def get_next_track_to_add(previous_artist, playlisttracks):
    if len(playlisttracks) == 0:
        return -1

    if _order == 0:
        return random.randrange(len(playlisttracks))
    elif _order == 2:
        for track in playlisttracks:
            if track['artist'][0] != previous_artist:
                return index
            index = index + 1
    if len(playlisttracks) > 0:
        return 0
    return -1


def add_tracks_to_playlist(artist, playlisttracks):
    index = 0
    previous_artist = artist
    while index < _maxcount or _maxcount < 0:
        i = get_next_track_to_add(previous_artist, playlisttracks)
        if i == -1:
            break
        previous_artist = playlisttracks[i]['artist']
        json_query = xbmc.executeJSONRPC(
            '{ "jsonrpc": "2.0", "method": "Playlist.Add", "params": { "playlistid": 0, "item": { "songid": %d } }, "id": 1 }' % playlisttracks[i]['songid'])
        json_query = unicode(json_query, 'ascii', errors='ignore')
        json_response = simplejson.loads(json_query)
        playlisttracks.pop(i)
        index = index + 1
    return index


if xbmc.Player().isPlayingAudio():
    tag = xbmc.Player().getMusicInfoTag()
    artist = tag.getArtist()
    title = tag.getTitle()
    playlist = xbmc.PlayList(0)
    currenttrackpos = playlist.getposition() + 1
    if currenttrackpos <= len(playlist):
        if _runinbackground:
            display_notification(_settings.get_string(1000), _settings.get_string(4000) % (artist.decode('ascii', 'ignore'), title.decode('ascii', 'ignore')))
        else:
            pDialog.update(25, _settings.get_string(3005), '%s - %s' % (artist.decode('ascii', 'ignore'), title.decode('ascii', 'ignore')))

        count, playlisttracks = get_similar_tracks(artist, title)
        log_debug('Found %d similar tracks in Kodi library' % count)

        if _runinbackground or not pDialog.iscanceled():
            index = 0
            if count > 0:
                while xbmc.PlayList(0).size() > currenttrackpos:
                    xbmc.PlayList(0).remove(xbmc.PlayList(0)[currenttrackpos].getfilename())
                index = add_tracks_to_playlist(artist, playlisttracks)

            if not _runinbackground:
                pDialog.close()

            log_debug('Added %d songs to playlist' % index)
            if _runinbackground:
                display_notification(_settings.get_string(1000), _settings.get_string(4001) % (index, artist.decode('ascii', 'ignore'), title.decode('ascii', 'ignore')))
            else:
                utils.ok(_settings.get_string(1000), _settings.get_string(3006) % index, '%s - %s' % (artist.decode('ascii', 'ignore'), title.decode('ascii', 'ignore')))
        else:
            log_debug('Script was cancelled')
    else:
        log_debug('Unable to get currently playing track')
else:
    utils.ok(_settings.get_string(1000), _settings.get_string(3008))
